### PR TITLE
[FIX] mass_mailing: allow to process queue as responsible


### PR DESCRIPTION
**Pull request by nle-odoo**
Add ir_config_parameter "mass_mailing.process_queue_with_responsible"
that when set will compute recipients and send the mail as the mass
mailing responsible (user_id) instead of superuser.

This is done to prevent the issue that by default, sending a mass
mailing to res.partner doesn't filter on company, so if you have access
to only one company with 20 users, you might think that the mass mailing
will only be sent to them, but possibly it could be sent to users of
other companies you don't know about.

With this changeset, we will use the responsible so the access rule
applying to the responsible will be used to restrict who will receive
the mass mailing.

opw-2532915

### CODE REVIEW

**Comment by nle-odoo** [comment]
todo: add test

**Comment by tde-banana-odoo** [comment]
Actually why using a parameter ? Seems using write_uid comes from an old fix ( https://github.com/odoo/odoo/commit/c3b49b3a5fec204a62c4a6741c6e7bf71c1bee55 ) that seems weird, looking at it. As statistics in master are sent using responsible user, shouldn't we directly user responsible to send mailings and avoid an unnecessary config parameter ?

**Comment by nle-odoo** [comment]
Hi @tde-banana-odoo 


We had two solutions in stable with odo:

- either add an ir.config_parameter, so people can enable it if they want (but they need to discover the issue)

- add a context flag in the cron ir_cron_mass_mailing_queue (this way this only impacts new installation, since we have noupdate=1)

Changing it for existing installations would mean that all of a sudden the company of the responsible:

- which with the new company switcher might be different than the one of the responsible,

- which could have changed between sending the mass mailing and that the mass mailing is sent

would all of a sudden be used.

**Comment by tde-banana-odoo** [comment]
> Hi @tde-banana-odoo
> 
> We had two solutions in stable with odo:
> 
> * either add an ir.config_parameter, so people can enable it if they want (but they need to discover the issue)

But nobody will be aware it exists... expect some specific people complaining probably.

> * add a context flag in the code of cron ir_cron_mass_mailing_queue (this way this only impacts new installation, since we have noupdate=1 on the cron)

I don't think adding "yet another flow breaking point" is a good idea :)

> 
> Changing it for existing installations would mean that all of a sudden the company of the responsible:
> 
> * which with the new company switcher might be different than the one of the responsible,

I am not sure to understand what you mean. But I don't understand why changing from write_uid to user_id changes anythong form company switcher point of view. Issue is the same for both people.

> * which could have changed between sending the mass mailing and that the mass mailing is sent

Then if your responsible changes, it makes more sense to send it based on the new responsible than based on the people who changed it (write_uit).

In short I don't really understand your answer. It seems to me we should always use user_id. If we feel this is a functional change, then I vote to do this change in 14 or master and do nothing in stable. Adding a config parameter nobody will be aware of is not really a good way to fix. Either we find something suitable, either we do nothing, according to me.

Cheers !

**Comment by nle-odoo** [comment]
Hi @tde-banana-odoo 

> I am not sure to understand what you mean. But I don't understand why changing from write_uid to user_id changes anythong form company switcher point of view. Issue is the same for both people.

The fix is about having:

```py
mass_mailing = mass_mailing.with_user(user)
```

so the evaluation of recipient filter is done with the responsible instead of the super user (which is not limited by access rule).

So if your mass mailing was sent to 1000 contacts of company A, B, C before doing the `with_user`, after doing it, it might be only sent to 5 contacts of company C.

The code could be changed to:

```diff
- user = mass_mailing.write_uid or self.env.user
+ user = mass_mailing.user_id
+ if force_responsible:
+     mass_mailing = mass_mailing.with_user(user)
```

without much impact if we want to remove the usage of write_uid.

**Comment by jakubPCV** [comment]
Hi guys, any update of the topic? This issue was created as a result of my ticket and it is actually a big problem as users with access to only one  company can easily by mistake sent mailing campaigns to all records belonging to all companies (even if not shared)
And I see for 15 days there has been no conclusion. Any estimation when this could make it into base Odoo code of v14?

**Comment by tde-banana-odoo** [comment]
> Hi @tde-banana-odoo
> 
> > I am not sure to understand what you mean. But I don't understand why changing from write_uid to user_id changes anythong form company switcher point of view. Issue is the same for both people.
> 
> The fix is about having:
> 
> ```python
> mass_mailing = mass_mailing.with_user(user)
> ```
> 
> so the evaluation of recipient filter is done with the responsible instead of the super user (which is not limited by access rule).
> 
> So if your mass mailing was sent to 1000 contacts of company A, B, C before doing the `with_user`, after doing it, it might be only sent to 5 contacts of company C.
> 
> The code could be changed to:
> 
> ```diff
> - user = mass_mailing.write_uid or self.env.user
> + user = mass_mailing.user_id
> + if force_responsible:
> +     mass_mailing = mass_mailing.with_user(user)
> ```
> 
> without much impact if we want to remove the usage of write_uid.

Actually I did not see that there was currently not any ``with_user``, just a strange ``with_context`` manipulation to fix a language issue. My bad. I thought we were just changing from write_uid to user_id, but actually admin is still sending mailings.

I still think right thing to do is the followup of https://github.com/odoo/odoo/pull/15467 : sending directly using a with_user based on responsible or write_uid or current user. It was evaluated at that time and default behavior should be to keep this domain evaluation. Doing this should also allow to remove the context manipulation if I am not wrong. Adding a config parameter is not really a solution as it does not solve issue for most people that will not be aware it exists.

Finally would love seeing some tests, especially due to language fix and domain evaluation that shoudl be tested once for all to avoid any future regression :) .

Thanks,

**Comment by C3POdoo** [comment]
Dear @nle-odoo,

Thank you for your contribution but the version 12.0 is no longer supported.
We only support the last 3 stable versions so no longer accepts patches into this branch.

We apology if we could not look at your request in time.
If the contribution still makes sense for the upper version, please let us know and do not hesitate to recreate one for the recent versions. We will try to check it as soon as possible.

_This is an automated message._
